providers/base/tests/test_network_device_info.py

```diff
@@ -0,0 +1,92 @@
+            self.assertEqual(addr, "***NOT CONFIGURED***")
+
+    def test_get_ipv4_address_with_connection(self):
+        test_input = (b'wlo1\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
+        b'\x00\x02\x00\x00\x00\xc0\xa8De\x00\x00\x00\x00\x00\x00\x00\x00'
+        b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
+        b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
+        b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
+        b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
+        b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
+        b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
+        b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
+        b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
+        b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
+        b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
+        b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
+        b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
+        b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
+        b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
+        b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00')
+        # test_input = b"I'm fake!!!"
+        mock_inet_ntoa = Mock(return_value=test_input)
+        with patch("fcntl.ioctl", mock_inet_ntoa):
+            interface = "wlo1"
+            addr = Utils.get_ipv4_address(interface)
+            self.assertEqual(addr, "192.168.68.101")
+
+class GetIpv6AddressTests(unittest.TestCase):
+    # There are 3 output we could get, but only 2 cases will happen due to the way we command it
+    # 1. No WiFi connected 
+    #    ""
+
+    # 2. Connected with WiFi
+    #    "2: wlo1    inet6 fe80::d9eb:3f93:c7b2:86ba/64 scope link noprefixroute \       valid_lft forever preferred_lft forever"
+
+    # 3. Connected with WiFi, but we ask the wrong interface name (This would not happen in our case, because the name is given by NetworkManager)
+    #    "Device "wlan0" does not exist."
+    def test_get_ipv6_address_without_connection(self):
+        test_input = ""
+        mock_check_output = Mock(return_value=test_input)
+        with patch("network_device_info.check_output", mock_check_output):
+            interface = "wlo1"
+            addr = Utils.get_ipv6_address(interface)
+            self.assertEqual(addr, "***NOT CONFIGURED***")
+
+    def test_get_ipv6_address_with_connection(self):
+        test_input = "2: wlo1    inet6 fe80::d9eb:3f93:c7b2:86ba/64 scope link noprefixroute \       valid_lft forever preferred_lft forever"
+        mock_check_output = Mock(return_value=test_input)
+        with patch("network_device_info.check_output", mock_check_output): # somehow it don't regonize check_output() if I don't put subprocess in front
```

Comment:
I'm not sure I understand the comment, since there is no `subprocess` in front of anything :)

But I think I know what you mean: mocking is hard! In this case, you need to point very specifically to what you want to mock. You don't want to mock the `check_output` method from the `subprocess` module, you want to mock the method being imported by `network_device_info`, hence you have to patch `network_device_info.check_output`. It's honestly not very clear...

A while ago, I listened to [this episode](https://talkpython.fm/episodes/show/287/testing-without-dependencies-mocking-in-python) of the "Talk Python to Me" podcast that helped me a lot to better understand testing and mocking. You can check it out.

Reply:
Ah sorry about that, this comment is me talking to myself while I trying to know how does mock works.
Thanks for link sharing, it really helps.